controller/engine_controller.go

```diff
@@ -411,19 +411,30 @@ func (ec *EngineController) DeleteInstance(obj interface{}) error {
 		return err
 	}
 
-	// Not assigned, safe to delete
+	var im *longhorn.InstanceManager
+	var err error
+	// Not assigned or not updated, try best to delete
 	if e.Status.InstanceManagerName == "" {
-		return nil
-	}
-
-	im, err := ec.ds.GetInstanceManager(e.Status.InstanceManagerName)
-	if err != nil {
-		if !apierrors.IsNotFound(err) {
-			return err
+		if e.Spec.NodeID == "" {
+			log.Warnf("Engine %v does not set instance manager name and node ID, will skip the actual process deletion", e.Name)
+			return nil
+		}
+		im, err = ec.ds.GetInstanceManagerByInstance(obj)
+		if err != nil {
+			log.Warnf("failed to detect instance manager for engine %v, will skip the actual process deletion: %v", e.Name, err)
```

Comment:
nit: do we have the conversion for the log first character should be upper case _or_ lower case?

Reply:
It's better to follow the conversion. Thanks for pointing it out! I have updated the PR.